fendermint/vm/interpreter/src/fvm/externs.rs

```diff
@@ -29,10 +63,60 @@ impl Consensus for FendermintExterns {
     }
 }
 
-impl Chain for FendermintExterns {
-    fn get_tipset_cid(&self, _epoch: ChainEpoch) -> anyhow::Result<Cid> {
-        unimplemented!("not expecting to use tipsets")
+impl<DB> Chain for FendermintExterns<DB>
+where
+    DB: Blockstore + Clone + 'static,
+{
+    // for retreiving the tipset_cid, we load the chain metadata actor state
+    // at the given state_root and retrieve the blockhash for the given epoch
+    fn get_tipset_cid(&self, epoch: ChainEpoch) -> anyhow::Result<Cid> {
+        // create a read only state tree from the state root
+        let bstore = ReadOnlyBlockstore::new(&self.blockstore);
+        let state_tree = StateTree::new_from_root(&bstore, &self.state_root)?;
+
+        // get the chain metadata actor state cid
+        let actor_state_cid = match state_tree.get_actor(CHAINMETADATA_ACTOR_ID) {
+            Ok(Some(actor_state)) => actor_state.state,
+            Ok(None) => {
+                return Err(anyhow!(
+                    "chain metadata actor id ({}) not found in state",
+                    CHAINMETADATA_ACTOR_ID
+                ));
+            }
+            Err(err) => {
+                return Err(anyhow!(
+                    "failed to get chain metadata actor ({}) state, error: {}",
+                    CHAINMETADATA_ACTOR_ID,
+                    err
+                ));
+            }
+        };
+
+        // get the chain metadata actor state from the blockstore
+        let actor_state: fendermint_actor_chainmetadata::State =
+            match state_tree.store().get_cbor(&actor_state_cid) {
+                Ok(Some(v)) => v,
+                Ok(None) => {
+                    return Err(anyhow!(
+                        "chain metadata actor ({}) state not found",
+                        CHAINMETADATA_ACTOR_ID
+                    ));
+                }
+                Err(err) => {
+                    return Err(anyhow!(
+                        "failed to get chain metadata actor ({}) state, error: {}",
+                        CHAINMETADATA_ACTOR_ID,
+                        err
+                    ));
+                }
+            };
+
+        match actor_state.get_block_hash(&bstore, epoch) {
+            Ok(Some(v)) => Ok(Cid::new_v1(DAG_CBOR, Code::Blake2b256.digest(&v))),
```

Comment:
```suggestion
            Ok(Some(v)) => Ok(Cid::new_v1(DAG_CBOR, Code::Blake2b256.wrap(&v))),
```

`.digest` would take the hash of the hash.

Reply:
Ahhh